File: algorithms/[338]比特位计数/solution.py
# topics = ["动态规划", "位运算"]

from typing import List


class Solution:
    def countBits(self, num: int) -> List[int]:
        """
        按位与运算（&）的一个性质是：对于任意整数 x，令 x=x&(x-1)，该运算将 x 的二进制表示的最后一个 1 变成 0
        """
        bits = [0] * (num + 1)
        for i in range(1, num + 1):
            bits[i] = bits[i & (i - 1)] + 1
        return bits

File: algorithms/[338]比特位计数/solution.ts
// topics = ["动态规划", "位运算"]

function countBits(num: number): number[] {
  const bits = new Array<number>(num + 1).fill(0);
  for (let i = 1; i <= num; i += 1) {
    // bits[x] 的值等于 bits[x/2] 的值加上 x 除以 2 的余数（即偶数不变，奇数加 1）
    bits[i] = bits[i >> 1] + (i & 1);
  }
  return bits;
}
